---
title: "Introduction to phyloGLM"
author: "Noah Dukler"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

First create some data and a tree. Pretend that sites are split into two classes, "Enhancer" and "Promoter" and also split by ontology classes "A" and "B".

```{r}
library(phyloGLM)
## Set seed for consistency
set.seed(123)
## Simulate random data
species=c("A","B","C","D","E")
aData=lapply(as.list(species),function(x) matrix(runif(120),ncol=3))
names(aData)=species

## Create test trees, both one that will fail, and one that will pass
tree=ape::rtree(n = length(species),tip.label = species)

## Create a data frame with labels
siteLabels=data.frame(cre.class=rep(c("Enhancer","Promoter"),each=nrow(aData[[1]])/2),
                      go.class=rep(c("A","B"),times=nrow(aData[[1]])/2))
```
Now create the allele data object:

```{r,echo=FALSE}
ad=alleleData(data=aData,tree=tree,siteInfo = siteLabels)
```

Now create a rate model specifying that genomic regions should be seperated by both cre.class and go.class.

```{r}
rateMod=rateModel(data = ad,rateFormula=formula(~cre.class+go.class))
```

Now we can fit the model:

```{r}
fitted_mod=fit(rateMod)
getParams(fitted_mod$model)
# failP=c(1.072339241,0.996833310,0.999776478,-2.463222270,-4.021448349,0.004552635,-6.568507110,-5.444297674,-4.300386918)
# rateMod@phylogeny$setParams(failP,1:length(failP)-1)
# siteLL(rateMod,40)
# zoo=numeric(40); for(i in 1:40) zoo[i]=rateMod@phylogeny$rate(0,rateMod@rateDM[i,,drop=FALSE])
# zooPi=matrix(nrow=40,ncol=3); for(i in 1:40) zooPi[i,]=rateMod@phylogeny$pi(rateMod@piDM[i,,drop=FALSE])
# for(i in 1:length(tree$edge.length)) print(rateMod@phylogeny$rateMatrix(zooPi[22,],zoo[22],tree$edge.length[i]))
```